File: Fundamentals/src/MidExam/ShoppingList.java
package MidExam;

import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public class ShoppingList {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);


        List<String> shoppingList = Arrays
                .stream(scanner.nextLine().split("\\!"))
                .collect(Collectors.toList());

        String input = scanner.nextLine();

        while (!input.equals("Go Shopping!")) {
            String[] command = input.split(" ");
            String commandLine = command[0];
            String product = command[1];


            switch (commandLine) {
                case "Urgent":
                    if (!shoppingList.contains(product)) {
                        shoppingList.add(0, product);
                    }
                    break;
                case "Unnecessary":
                    if (shoppingList.contains(product)) {
                        shoppingList.remove(product);
                    }
                    break;
                case "Correct":
                    String newItem = command[2];
                    if (shoppingList.contains(product)) {
                        int index = shoppingList.indexOf(product);
                        shoppingList.remove(product);
                        shoppingList.add(index, newItem);

                    }

                    break;
                case "Rearrange":
                    if (shoppingList.contains(product)) {
                        shoppingList.remove(product);
                        shoppingList.add(product);

                    }

                    break;
            }


            input = scanner.nextLine();
        }

        System.out.println(String.join(", ", shoppingList));
    }


}

File: JS Front-End/Exam preparation/Shopping List.js
function solve(input){
    let arr = input.shift().split('!');
    console.log(arr);
    

    for (const line of input) {
        if(line === 'Go Shopping!') {
            break;
        }

        let data = line.split(' ');
        let command = data[0];
        let commands = [...data.slice(1)];
        
        if (command === 'Urgent'){
            let [item] = commands;
            if (!arr.includes(item)){
                arr.unshift(item);
            }

        }else if(command === 'Unnecessary'){
            let [item] = commands;
            if (arr.includes(item)){
                let index = arr.indexOf(item);
                arr.splice(index,1);
            }
        }else if(command === 'Correct'){
            let [item, newItem] = commands;
            if (arr.includes(item)){
             let index = arr.indexOf(item);
            arr[index] = newItem;
            }
        }else if(command === 'Rearrange'){
            let [item] = commands;
            if (arr.includes(item)){
                let index = arr.indexOf(item);
                arr.splice(index,1);
                arr.push(item);
            }
        }

    }

    console.log(arr.join(', '));



}
solve((["Tomatoes!Potatoes!Bread",
"Unnecessary Milk",
"Urgent Tomatoes",
"Go Shopping!"])
)
solve((["Milk!Pepper!Salt!Water!Banana",
"Urgent Salt",
"Unnecessary Grapes",
"Correct Pepper Onion",
"Rearrange Grapes",
"Correct Tomatoes Potatoes",
"Go Shopping!"])
)
solve((["Tomatoes!Potatoes!Bread",
"Unnecessary Milk",
"Correct Tomatoes Onion",
 "Urgent Milk",
 "Unnecessary Milk",
"Go Shopping!"])
)
